modules/Router.js

```diff
@@ -1,219 +1,185 @@
-                if (error) {
-                  callback(error);
-                } else {
-                  nextState.components = components;
-                  callback(null, nextState);
-                }
-              });
-            }
-          });
-        }
-      });
-    }
-
-  },
-
-  childContextTypes: {
-    router: object
-  },
-
-  getChildContext() {
-    return {
-      router: this
-    };
-  },
-
-  propTypes: {
-    createElement: func,
-    parseQueryString: func,
-    onError: func,
-    onUpdate: func,
-    routes,
+import createMemoryHistory from 'history/lib/createMemoryHistory';
+import useQueries from 'history/lib/useQueries';
+import matchUntilResolved from './matchUntilResolved';
+import { pick, inArray } from './CollectionUtils';
+
+function noop() {}
+
+const createFallbackHistory = createMemoryHistory
+
+const componentSpecificOptions = [
+  'createElement'
+];
+
+export default class Router extends Component {
+  static propTypes = {
+    createElement: PropTypes.func,
+    parseQueryString: PropTypes.func,
+    onError: PropTypes.func,
+    onUpdate: PropTypes.func,
+    routes: RouterPropTypes.routes,
     // Routes may also be given as children (JSX)
-    children: routes,
-
-    // Client
-    history,
+    children: RouterPropTypes.routes,
 
     // Unit testing, simple server
-    location,
+    location: RouterPropTypes.location,
 
     // Flux, data fetching
-    initialState: object
-  },
+    // @deprecated, use <RouterRenderer {...initialState} /> instead
+    initialState: PropTypes.object
+  }
 
-  getDefaultProps() {
-    return {
-      createElement,
-      parseQueryString
-    };
-  },
+  static defaultProps = {
+    parseQueryString,
+    onError: error => { throw error; },
+    onUpdate: noop
+  }
+
+  /**
+   * Server-side shortcut for `router.match()`. Use `router.match()`
+   * directly instead.
+   * @deprecated
+   * @param {Array<Route>} routes
+   * @param {Location} location
+   * @param {Function} callback
+   * @param {Object} [initialState]
+   */
+  static run(routes, location, callback, initialState) {
+    const router = createReactRouter(createFallbackHistory)({
+      routes,
+      initialState
+    });
+    router.match(initialState, location, callback);
+  }
 
-  getInitialState() {
-    return {
+  constructor(props, context) {
+    super(props, context);
+    const {
+      history,
+      routes,
+      children
+    } = props;
+
+    this.state = {
       isTransitioning: false,
-      location: null,
-      routes: null,
-      params: null,
-      components: null,
-      ...this.props.initialState
     };
-  },
 
-  updateLocation(location) {
-    if (!location.query)
-      location.query = this.props.parseQueryString(location.search.substring(1));
+    const routerProps = pick(props, k => !inArray(componentSpecificOptions, k));
 
-    this.setState({
-      isTransitioning: true
-    });
+    const createHistory = history
+      ? () => history
```

Comment:
We probably don't have tests for this, but this line will ignore the `parseQueryString` prop. Ideally this would not be a `<Router>` prop, but would instead be configured when someone creates their `history` object. But I'd still like to support it at this level.